src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+    d = material.drivers
+
+    theta0 = p.theta0
+    Ef = p.E
+    nuf = p.nu
+    alphaf = p.alpha
+
+    temperature = d.temperature
+    dstrain = dd.strain
+    dtime = dd.time
+    dtemperature = dd.temperature
+    @unpack stress, X1, X2, X3, plastic_strain, cumeq, R = v
+
+    VariableState{U} = GenericChabocheThermalVariableState{U}
+    DriverState{U} = GenericChabocheThermalDriverState{U}
+    ff(sigma, R, X1, X2, X3, theta) = yield_criterion(VariableState{T}(stress=sigma, R=R, X1=X1, X2=X2, X3=X3),
+                                                      DriverState{T}(temperature=theta),
+                                                      p)
+    # n = ∂f/∂σ
+    nf(sigma, R, X1, X2, X3, theta) = yield_jacobian(VariableState{T}(stress=sigma, R=R, X1=X1, X2=X2, X3=X3),
+                                                     DriverState{T}(temperature=theta),
+                                                     p)
+    # p'  (dp = p' * dtime)
+    dotpf(sigma, R, X1, X2, X3, theta) = overstress_function(VariableState{T}(stress=sigma, R=R, X1=X1, X2=X2, X3=X3),
+                                                             DriverState{T}(temperature=theta),
+                                                             p)
+
+    # Compute the elastic trial stress.
+    #
+    # We compute the elastic trial stress increment by using data from the
+    # start of the timestep, so we have essentially a forward Euler predictor.
+    #
+    # Relevant equations (thermoelasto(-visco-)plastic model):
+    #
+    #   ε = εel + εpl + εth
+    #   σ = D : εel           (Hooke's law)
+    #
+    # where D = D(θ) is the elastic stiffness tensor (symmetric, rank-4),
+    # and θ is the absolute temperature (scalar, θ > 0).
+    #
+    # Thus:
+    #
+    #   Δσ = Δ(D : εel)
+    #      = ΔD : εel + D : Δεel
+    #      = (dD/dθ Δθ) : εel + D : Δεel
+    #      = dD/dθ : εel Δθ + D : Δεel
+    #
+    # where the elastic strain increment
+    #
+    #   Δεel = Δε - Δεpl - Δεth
+    #
+    # In the elastic trial step, we temporarily assume Δεpl = 0, so then:
+    #
+    #   Δεel = Δε - Δεth
+    #
+    # The elastic stiffness tensor D is explicitly known. Its derivative dD/dθ
+    # we can obtain by autodiff. The temperature increment Δθ is a driver.
+    #
+    # What remains to consider are the various strains. Because we store the total
+    # stress σ, we can obtain the elastic strain εel by inverting Hooke's law:
+    #
+    #   εel = C : σ
+    #
+    # where C = C(θ) is the elastic compliance tensor (i.e. the inverse of
+    # the elastic stiffness tensor D with respect to the double contraction),
+    # and σ is a known total stress. (We have it at the start of the timestep.)
+    #
+    # The total strain increment Δε is a driver. So we only need to obtain Δεth.
+    # The thermal strain εth is, generally speaking,
+    #
+    #   εth = α(θ) (θ - θ₀)
+    #
+    # where α is the linear thermal expansion tensor (symmetric, rank-2), and
```

Comment:
I am more used to the form
εth = α(θ) (θ - θ₀) I,
where α is a scalar and the unit tensor I is shown explicitly :) 
Does not need to be changed, as long as we are consistent with this notation.

Reply:
I'm fine with either form.

Yes, the notation conflict is unfortunate. Let's change the symbol.

If I recall, α is commonly used also for the thermal expansion coefficient. So we don't have a good single-letter candidate; maybe we need something clunky like `thermalalpha`.

Reply:
Yes &alpha; is used commonly for thermal expansion coefficient. Maybe alpha_th or something could work, or then we could have a systematic notation for tensor-variables versus scalars.